### DISCUSSION TITLE
Interface Requirements instead abstract static interface members

### DISCUSSION THREAD

**Discussion by jv-np** [dotnet/csharplang#6035] [Language Ideas] [answered]
Why instead abstract interface members, we do not interface requirements?

My argument is:
 * It provides far more flexibility
 * It can provide optimization avoiding boxing and thus not generating garbage
 * It can ensure that code require only code it will actually consume

So I propose interfaces with requirement-only side of contracts.
So, to be more clear, what I mean with interfaces with requirement-only side of contracts are interfaces that can be defined but need not to be implemented by types, instead, the compiler will just ensure that the specified type matches the implementation and generate apropriate code when calling methods and using properties for static and instance. This way C# can remove boxing for interfaces too on types, so, if abstract interfaces was designed to principally implement math operators this make far much more sense as it will be used principally with value types. The result of an implementation like that will be like:

```csharp
public requirements interface IAdd
{
  static IAdd operator + (IAdd left, IAdd right);
} 
...
public class ConsumeIAdd
{
  public T SumThreeValues<T>(T a, T b, T c) where T : IAdd
 {
    return a + b + c;
 }
}
```
The code above should work simply without needing to make a type inherit from IAdd, just like a type constraint that will be applied.

This solves the problem of needing to update all types with math interfaces in order to use math operators in methods and open the possibility to remove boxing wich can be a problem in some scenarious. This idea can solve some of Linq problems too if it is implemented in certain ways, like making an IReqEnumerable<T> that will be used instead IEnumerable<T> and thus removing the requirement to create boxes and garbage when iterating over collections.

Other way to implement this would be to just add a constraint called 'require', 'formally', 'compromise' or something similar that will make compiler need not to look if type implemented interface but instead if type is implementing the interface formal requirements by signatures. This version can be used with abstract interface members and would be like:
```csharp
public interface IAdd<T>
{
 static abstract T operator +(T left, T right);
}
...
public class ConsumeIAdd
{
 public T SumThreeValues<T>(T a, T b, T c) where T : require IAdd
 {
    return a + b + c;
 }
}
```
As types need not to implement interfaces in this version, this would not be a breaking change and thus is preferable over normal abstract static interface members approach to problem.

**Comment by CyrusNajmabadi** [comment, marked as answer]
@JoaoVictorVP that idea is already being discussed in the shapes/roles/extensions designs.

**Reply by jv-np** [reply]
Thank you my friend

**Reply by jv-np** [reply]
You're sure the roles/extensions proposal are like mine?

**Reply by CyrusNajmabadi** [reply]
More or less.  It allows mapping an existing type to support a particular 'shape' without the type needing to explicitly implement it itself.  That mapping can be done after the fact for the entire type, or even just particular instances of that types.
